File: kotlin_project/lbsolverlib/src/main/kotlin/com/tinybitsinteractive/lbsolverlib/Solver.kt
package com.tinybitsinteractive.lbsolverlib

import java.net.HttpURLConnection
import java.net.URL
import java.nio.file.Path
import kotlin.io.path.deleteIfExists
import kotlin.io.path.div
import kotlin.io.path.exists
import kotlin.io.path.writeLines
import kotlin.time.Duration
import kotlin.time.measureTime
import kotlin.time.measureTimedValue

import com.tinybitsinteractive.lbsolverlib.kotlincore.KotlinCore
import com.tinybitsinteractive.lbsolverlib.nativecore.NativeCore

typealias CompletionHandler = (String?, elapsed: Duration?) -> Unit

class Solver {
    private var error: String? = null
    private var kotlinCore: SolverCore? = KotlinCore()
    private var nativeCore: SolverCore? = NativeCore()
    private val logger: Logger by lazy {
        Logger.factory.create("Solver")
    }

    enum class Type {
        Kotlin,
        Native
    }

    companion object {
        fun cleanSides(sides: String): String {
            var cleaned = ""
            var liveCount = 0
            for (c in sides) {
                val lc = c.lowercase()[0]
                if (lc.isLetter() && lc !in cleaned) {
                    cleaned += lc
                    ++liveCount
                    if (liveCount == 12) {
                        break
                    }
                    if ((liveCount % 3) == 0) {
                        cleaned += ' '
                    }
                }
            }
            return cleaned.trim()
        }
    }

    fun setup(cachePath: Path, onReady: CompletionHandler) {
        logger.info("setup()")
        if (kotlinCore == null) {
            logger.err("destroy() has been called.")
            onReady("destroy() has been called.", null)
            return
        }
        Thread {
            val wordsPath = cachePath / wordsUrl.substring(wordsUrl.indexOfLast{ it == '/'} + 1)
            val setupTime = measureTime {
                if (error == null) {
                    error = fetchWords(wordsPath)
                }
                if (error == null) {
                    error = kotlinCore?.setup(cachePath, wordsPath)
                }
                if (error == null) {
                    error = nativeCore?.setup(cachePath, wordsPath)
                }
            }
            onReady(error, setupTime)
        }.start()
    }

    fun destroy() {
        kotlinCore?.destroy()
        kotlinCore = null
        nativeCore?.destroy()
        nativeCore = null
    }

    fun getError() = error

    fun solve(type: Type, box: String, onComplete: CompletionHandler) {
        logger.info("solve($type, $box)")

        if (kotlinCore == null) {
            logger.err("destroy() has been called.")
            onComplete("destroy() has been called.", null)
            return
        }

        if (box != cleanSides(box) || box.length != 15) {
            onComplete("$box is not a valid puzzle", null)
            return
        }

        val core = if (type == Type.Native) nativeCore!! else kotlinCore!!

        Thread {
            val (solutions, elapsed) = measureTimedValue {
                core.solve(box)
            }
            onComplete(solutions, elapsed)
        }.start()
    }

    private fun fetchWords(destPath: Path): String? {
        if (!destPath.exists()) {
            return null
        }
        try {
            with(URL(wordsUrl).openConnection() as HttpURLConnection) {
                inputStream.bufferedReader().use { br ->
                    destPath.writeLines(br.readLines())
                }
            }
        } catch (_: Throwable) {
            destPath.deleteIfExists()
            return "failed downloading words."
        }
        return null
    }
}

val Solver.Companion.wordsUrl: String
  get() = "https://raw.githubusercontent.com/benn-herrera/letterboxed/refs/heads/main/src/letterboxed/words_alpha.txt"

File: swift_project/lbsolverlib/Sources/lbsolverlib/Solver.swift
import Foundation

public enum SolverType {
    case Swift
    case Native
}

public typealias CompletionHandler = (String?, TimeInterval) -> Void

public class Solver {
    public static func clean_puzzle(_ pz: String) -> String {
        var pz_clean = ""
        var live_count = 0
        for c in pz.lowercased() {
            if (pz_clean.contains(c) || !c.isLetter) {
                continue
            }
            live_count += 1
            pz_clean.append(c)
            if (live_count == 12) {
                break
            }
            if (live_count % 3) == 0 {
                pz_clean.append(" ")
            }
        }
        return pz_clean
    }
    
    public init() {}
    
    public func get_error() -> String? {
        return error
    }
    
    public func setup(cache_path: URL, on_ready: @escaping CompletionHandler) {
        Thread.detachNewThread {
            let words_path = URL(fileURLWithPath: cache_path.path() + "/words_alpha.txt")
            let start = Date()
            let swift_core = SwiftCore()
            if let sc_err = swift_core.setup(cache_path: cache_path, words_path: words_path) {
                self.error = self.error != nil ? (self.error! + "\n" + sc_err) : sc_err
            }
            else {
                self.swift_core = swift_core
            }
            let native_core = NativeCore()
            if let nc_err = native_core.setup(cache_path: cache_path, words_path: words_path) {
                self.error = self.error != nil ? (self.error! + "\n" + nc_err) : nc_err
            }
            else {
                self.native_core = native_core
            }
            let elapsed = Date().timeIntervalSince(start)
            on_ready(self.error, elapsed)
        }
    }
    
    public func solve(_ puzzle: String, type: SolverType, on_complete: @escaping CompletionHandler) {
        Thread.detachNewThread {
            guard let core = (type == .Native ? self.native_core : self.swift_core) else {
                on_complete("error: \(type) solver core not set up", 0)
                return
            }
            let start = Date()
            let solutions = core.solve(puzzle: puzzle)
            let elapsed = Date().timeIntervalSince(start)
            on_complete(solutions, elapsed)
        }
    }
    
    private var swift_core: SolverCore? = nil
    private var native_core: SolverCore? = nil
    private var error: String? = nil

    private static let words_url =         "https://raw.githubusercontent.com/benn-herrera/letterboxed/refs/heads/main/src/letterboxed/words_alpha.txt"
}
